src/iop/lut3d.c

```diff
@@ -0,0 +1,1010 @@
+    if(clut_cl == NULL)
+    {
+      fprintf(stderr, "[lut3d process_cl] error allocating memory\n");
+      err = CL_MEM_OBJECT_ALLOCATION_FAILURE;
+      goto cleanup;
+    }
+    if (transform)
+    {
+      const int success = dt_ioppr_transform_image_colorspace_rgb_cl(devid, dev_in, dev_out, width, height, work_profile, lut_profile, "linrec2020 to LUT Space");
+      if (!success)
+       transform = FALSE;
+    }
+    size_t sizes[] = { ROUNDUPWD(width), ROUNDUPHT(height), 1 };
+    if (transform)
+      dt_opencl_set_kernel_arg(devid, kernel, 0, sizeof(cl_mem), (void *)&dev_out);
+    else
+      dt_opencl_set_kernel_arg(devid, kernel, 0, sizeof(cl_mem), (void *)&dev_in);
+    dt_opencl_set_kernel_arg(devid, kernel, 1, sizeof(cl_mem), (void *)&dev_out);
+    dt_opencl_set_kernel_arg(devid, kernel, 2, sizeof(int), (void *)&width);
+    dt_opencl_set_kernel_arg(devid, kernel, 3, sizeof(int), (void *)&height);
+    dt_opencl_set_kernel_arg(devid, kernel, 4, sizeof(cl_mem), (void *)&clut_cl);
+    dt_opencl_set_kernel_arg(devid, kernel, 5, sizeof(int), (void *)&level);
+    err = dt_opencl_enqueue_kernel_2d(devid, kernel, sizes);
+    if (transform)
+      dt_ioppr_transform_image_colorspace_rgb_cl(devid, dev_out, dev_out, width, height, lut_profile, work_profile, "LUT Space to linrec2020");
+  }
+  else  // no clut, do nothing
+  {
+    err = CL_INVALID_VALUE;
+  }
+  if(err != CL_SUCCESS)
+  {
+    fprintf(stderr, "[lut3d process_cl] error %i enqueue kernel\n", err);
+    goto cleanup;
+  }
+
+cleanup:
+  if(clut_cl) dt_opencl_release_mem_object(clut_cl);
+
+  if(err != CL_SUCCESS) dt_print(DT_DEBUG_OPENCL, "[opencl_lut3d] couldn't enqueue kernel! %d\n", err);
+  return (err == CL_SUCCESS) ? TRUE : FALSE;
+}
+#endif
+
+void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const void *const ibuf, void *const obuf,
+             const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+  dt_iop_lut3d_data_t *d = (dt_iop_lut3d_data_t *)piece->data;
+  const int width = roi_in->width;
+  const int height = roi_in->height;  
+  const int ch = piece->colors;
+  const float *const clut = (float *)d->clut;
+  const uint8_t level = d->level;
+  const dt_interpolation_worker interpolation_worker
+    = (d->params.interpolation == DT_IOP_TETRAHEDRAL) ? correct_pixel_tetrahedral
+    : (d->params.interpolation == DT_IOP_TRILINEAR) ? correct_pixel_trilinear
+    : correct_pixel_pyramid;
+  const int colorspace
+    = (d->params.colorspace == DT_IOP_SRGB) ? DT_COLORSPACE_SRGB
+    : (d->params.colorspace == DT_IOP_REC709) ? DT_COLORSPACE_REC709
+    : (d->params.colorspace == DT_IOP_LIN_REC709) ? DT_COLORSPACE_LIN_REC709
+    : DT_COLORSPACE_LIN_REC2020;
+  const dt_iop_order_iccprofile_info_t *const lut_profile
+    = (colorspace != DT_COLORSPACE_LIN_REC2020) ? dt_ioppr_add_profile_info_to_list(self->dev, colorspace, "", INTENT_PERCEPTUAL) : NULL;

```

Comment:
Why lut_profile is NULL if colorspace != DT_COLORSPACE_LIN_REC2020?

Reply:
Because DT_COLORSPACE_LIN_REC2020 is the default color space for the module, so we we skip the transform.

Reply:
The input image may not be in lin rec 2020, that's up to the user.

Reply:
I don't understand the point.
The input image can be in whatever color space, sure.
As the module requests the lin rec 2020 as default color space (which works between colorin and colorout) it get the image in in lin rec 2020. So if the lut color space is also lin rec 2020 there is no need to transform it. If the lut color space is different then the transformation is done.
Otherwise, before colorin, the module receives the image in raw rgb and after colorout in sRGB, without any option to change it.
Where am I wrong ?

Reply:
You can't ask for a particular rgb color space, the input image will always be in the work color space. If you want it to be in any particular rgb color space you have to transform it.

Reply:
I see. I've missed that input color profile could change the the work profile.
So I have to check if the work profile is the required one for LUT or not before making the transformation. Correct ?

Reply:
The dt_ioppr_transform_image_colorspace_rgb() already does that, so you can call it if both profiles are not NULL, or you can do the check before the call if it gives you better performance.

Reply:
Great. Thanks. I let dt_ioppr_transform_image_colorspace_rgb() do the job then.